rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+            attention_weights: 5-D tensor containing self-attention weights received 
+                from the underlying transformer. `None` if the transformer has 0 layers.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        mask_sequence = inputs[2]
+        mask_sequence_sentence = inputs[3]
+
+        # Combine all features (sparse/dense, sequence-/sentence-level) into one tensor
+        seq_sent_features = self._tf_layers["input_layer"](
+            (
+                sequence_features,
+                sentence_features,
+                mask_sequence,
+                mask_sequence_sentence,
+            )
+        )
+
+        seq_sent_features = self._tf_layers["ffnn"](seq_sent_features, training)
+
+        # Produce helper variables for masked language modelling (only in training mode)
+        if self.enables_mlm and training:
+            # token_ids are embeddings with the same shape as the first dense sequence
+            # feature or alternatively the last dimension will have size 2 if there are
+            # only sparse sequence features present. In any case, these embeddings can
+            # be viewed as labels/unique IDs of all input tokens (to be used later)
+            # because these embeddings aren't affected by dropout or masking and are
+            # effectively always unique for different input tokens (and same for the
+            # same tokens).
+            token_ids = self._features_as_token_ids(sequence_features)
+            # Pad in the sequence dimension to match the shape of combined sequence- and
+            # sentence-level features (sentence-level features effectively have sequence
+            # length of 1).
+            token_ids = tf.pad(token_ids, [[0, 0], [0, 1], [0, 0]])
+
+            # mlm_mask_bool has the same shape as mask_sequence_sentence (i.e. as the
+            # tensor with all combined features), with True meaning tokens that are
+            # masked and False meaning tokens that aren't masked or that are fake
+            # (padded) tokens.
+            transformer_inputs, mlm_mask_bool = self._tf_layers["mlm_input_mask"](
+                seq_sent_features, mask_sequence_sentence, training
+            )
+        else:
+            token_ids = None
+            transformer_inputs = seq_sent_features
+            mlm_mask_bool = None
+
+        # Apply the transformer (if present), hence reducing a sequences of features per
+        # input example into a simple fixed-size embeddings.
+        if "transformer" in self._tf_layers:
+            mask_padding = 1 - mask_sequence_sentence
+            outputs, attention_weights = self._tf_layers["transformer"](
+                transformer_inputs, 1 - mask_padding, training
+            )
+            outputs = tfa.activations.gelu(outputs)
+        else:
+            outputs, attention_weights = transformer_inputs, None
```

Comment:
I don't think returning `None` is a possibility in `keras.call`?

Reply:
Apparently, it works. I think the big difference is between returning just a `None`, or effectively a tuple of multiple outputs, one of them being `None`. (Perhaps it's similar to the inputs to a Keras layer -- you can't have the first argument be `[]` but you can pass a tuple containing also a `[]` or `None` element.)

Reply:
it is a bit of a stretch, but are you sure it doesn't create multiple graphs in this case? could be a reason for the OOM

Reply:
Of course I'm not 100% sure, though I think this is not the reason for OOM. By the way, the relevant Keras code is [here](https://github.com/tensorflow/tensorflow/blob/4f6924d44aa914dbfee79fee8211bdbc25c6e089/tensorflow/python/keras/engine/base_layer.py#L1113). It only checks for the entire output being `None` (though the message says that the output should be a Tensor or a list of Tensors, so probably they're not super happy if the list contains `None`...).

If you were to change the code, would you replace the `None`s with `tf.zeros((0,))`? Personally, I'd try to avoid not returning the `None` things at all, because then you have to place `if`s wherever you call the layer in order to expect the correct number of outputs (depending on whether this layer internally has a transformer or not)...

Reply:
> Personally, I'd try to avoid not returning the None things at all
 what do you mean? your function always should return a fixed number of outputs

Reply:
the sake of consistency I'd return the `tf.zeros((0,))` since that's what we do in monster ted

Reply:
> your function always should return a fixed number of outputs

I think I've seen places where this is not the case, that's why I wanted to mention it as a possible alternative. But I'm glad that we both disagree with it 😄

Reply:
> I think I've seen places where this is not the case

in rasa?

Reply:
Alright, this should now be addressed 🙂

Reply:
@Ghostvv I'm really not sure, sorry. May have been in Rasa or elsewhere.